orchagent/cbforch.cpp

```diff
@@ -0,0 +1,322 @@
+
+                if (sai_oid != SAI_NULL_OBJECT_ID)
+                {
+                    SWSS_LOG_INFO("Successfully created %s map", getMapName());
+                    emplace(map_id, sai_oid);
+                    success = true;
+                }
+                else
+                {
+                    SWSS_LOG_ERROR("Failed to create %s map %s",
+                                    getMapName(),
+                                    map_id.c_str());
+                    success = false;
+                }
+            }
+            else
+            {
+                SWSS_LOG_NOTICE("Updating existing %s map %s",
+                                 getMapName(),
+                                 map_id.c_str());
+
+                success = updateMap(map_it->second, map_list);
+
+                if (success)
+                {
+                    SWSS_LOG_INFO("Successfully updated %s map", getMapName());
+                }
+                else
+                {
+                    SWSS_LOG_ERROR("Failed to update %s map %s",
+                                    getMapName(),
+                                    map_id.c_str());
+                }
+            }
+
+            // Remove the allocated list
+            delete[] map_list.list;
+        }
+        else if (op == DEL_COMMAND)
+        {
+            if (map_it != end())
+            {
+                SWSS_LOG_NOTICE("Deleting %s map %s",
+                                 getMapName(),
+                                 map_id.c_str());
+
+                success = removeMap(map_it->second);
+
+                if (success)
+                {
+                    SWSS_LOG_INFO("Successfully removed %s map", getMapName());
+                    erase(map_it);
+                }
+                else
+                {
+                    SWSS_LOG_ERROR("Failed to remove %s map %s",
+                                    getMapName(),
+                                    map_id.c_str());
+                }
+            }
+            else
+            {
+                SWSS_LOG_WARN("Tried to delete inexistent %s map %s",
+                               getMapName(),
+                               map_id.c_str());
+                // Mark it as success to remove it from the consumer
+                success = true;
+            }
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s", op.c_str());
+            // Set success to true to remove the operation from the consumer
+            success = true;
+        }
+
+        if (success)
+        {
+            SWSS_LOG_DEBUG("Removing consumer item");
+            it = consumer.m_toSync.erase(it);
+        }
+        else
+        {
+            SWSS_LOG_DEBUG("Keeping consumer item");
+            ++it;
+        }
+    }
+}
+
+/*
+ * Purpose:     Extract the QoS map from the Redis's field-value pairs.
+ *
+ * Description: Iterate over the field-value pairs and create a SAI QoS map.
```

Comment:
The description here should make it clear that this function allocates memory, and that it is the responsibility of the caller to free off the returned map_list's .list field memory after it's finished using it.